.buildkite/test-scenarios.sh

```diff
@@ -0,0 +1,152 @@
+create-scenario-step() {
+  local scenario=$1
+  local label=$2
+  local image=${3:-test}
+
+  # Output the Buildkite step for building this scenario using the step template
+  # defined in the base template file.
+  cat <<YAML
+  - <<: *step-defaults
+    <<: *docker-agents
+    label: ":node: ${label}"
+    key: "${image}_${scenario//-/_}"
+    depends_on: "${image}_image"
+    artifact_paths:
+      - .buildkite/artifacts/**/*
+    env:
+      SCENARIO: ${scenario}
+    command: yo --no-insight web-starter --force --debug --uninteractive
+    plugins:
+      - *ecr
+      - ${DOCKER_COMPOSE_PLUGIN}:
+          run: "${image}"
+          debug: true
+      - improbable-eng/metahook:
+          pre-artifact: .buildkite/tar-generator-output.sh $image $scenario
+
+YAML
+}
+
+# Usage: create-compare-step <scenario_name> <scenario_label> <base_image> <target_image>
+# * scenario_name is a directory name at test/<scenario_name>
+# * scenario_label is a readable label to display for this test scenario
+# * base_image (optional) is the base image to compare against. Defaults to test
+# * target_image (optional) is the base image to compare against. Defaults to stable
+create-compare-step() {
+  local scenario=$1
+  local label=$2
+  local image_1=${3:-test}
+  local image_2=${4:-stable}
+
+  # Output the Buildkite step for building this scenario using the step template
+  # defined in the base template file.
+  cat <<YAML
+
+  - label: ':mag: Compare: ${label}'
+    <<: *step-defaults
+    <<: *docker-agents
+    depends_on:
+      - '${image_1}_${scenario//-/_}'
+      - '${image_2}_${scenario//-/_}'
+    env:
+      SCENARIO: ${scenario}
+    plugins:
+      # Download the tarball into the test directory for use building the Docker image.
+      - artifacts#v1.3.0:
+          download: ".buildkite/artifacts/**/${scenario}/${scenario}.tgz"
+          upload: ".buildkite/artifacts/${scenario}.html"
+      - docker-compose#v3.7.0:
+          run: diff
+          debug: true
+          workdir: '/artifacts'
+          # Run as the root user since Buildkite's filesystem sets root ownership of files by default.
+          user: root
+          # Disable shell to pass commands directly into the image's entrypoint script.
+          shell: false
+          command:
+            - --exclude-directory-metadata=recursive
+            - --no-default-limits
+            - --new-file
+            - --exclude
+            - "${scenario}.tgz"
+            - --html
+            - /artifacts/${scenario}.html
+            - ${image_1}/${scenario}
+            - ${image_2}/${scenario}
+          volumes:
+            - './.buildkite/artifacts:/artifacts'
+      - improbable-eng/metahook:
+          pre-command: |
+            set -ex
+            # Extract artifact gzips for direct file comparisons.
+            $(cd ".buildkite/artifacts/${image_1}/${scenario}" && tar -xf "${scenario}.tgz")
```

Comment:
The `$(...)` syntax captures and returns output. In this case, since you have it in the command position, the shell will attempt to execute the output. It's easy to mix up with subshells (which are just `(...)` - no leading `$`).

I suggest just using this one-liner, adding the `-v` (verbose) flag to tar instead of executing a subsequent `ls`:

```sh
(cd ".buildkite/artifacts/${image_1)/${scenario}" && tar -xvf "${scenario}.tgz")
```